File: jwt_bruteforce.py
import base64
import hashlib
import hmac
import json
import sys
import threading
import time

# Global state
running = True
candidate = ""
start_time = 0

# Charset: common characters used in industry
CHARSET = list(
    "abcdefghijklmnopqrstuvwxyz"
    "ABCDEFGHIJKLMNOPQRSTUVWXYZ"
    "0123456789"
    "!@#$%^&*()-_+=[]{};:'\",.<>?/\\|`~"
)


def extract_alg_from_header(header_json_str):
    try:
        header = json.loads(header_json_str)
        return header.get("alg")
    except json.JSONDecodeError:
        return None


def hmac_base64url(data, key, alg):
    if alg == 'HS256':
        digest_mod = hashlib.sha256
    elif alg == 'HS384':
        digest_mod = hashlib.sha384
    elif alg == 'HS512':
        digest_mod = hashlib.sha512
    else:
        raise ValueError(f"Unsupported alg: {alg}")

    hm = hmac.new(key.encode('utf-8'), data.encode('utf-8'), digest_mod)
    sig = base64.urlsafe_b64encode(hm.digest()).decode('utf-8').rstrip('=')
    return sig


def timer_thread():
    global running, candidate, start_time
    while running:
        elapsed = time.time() - start_time
        hours = int(elapsed // 3600)
        minutes = int((elapsed % 3600) // 60)
        seconds = int(elapsed % 60)
        hundredths = int((elapsed - int(elapsed)) * 100)
        sys.stdout.write(
            f"⏱  Trying      : {candidate:<30} Time elapsed  : {hours:03d}:{minutes:02d}:{seconds:02d}.{hundredths:02d}\r"
        )
        sys.stdout.flush()
        time.sleep(0.1)


def brute_force(signing_input, target_sig, alg, length, pos, current):
    global candidate
    if pos == length:
        candidate = ''.join(current)
        test_sig = hmac_base64url(signing_input, candidate, alg)
        return test_sig == target_sig
    for c in CHARSET:
        current[pos] = c
        if brute_force(signing_input, target_sig, alg, length, pos + 1, current):
            return True
    return False


def main():
    global running, start_time
    jwt = input("Enter JWT token\t		: ").strip()
    try:
        max_length = int(input("Enter max secret length to try\t: ").strip())
    except ValueError:
        print("Invalid length input.")
        return

    parts = jwt.split('.')
    if len(parts) != 3:
        print("Invalid JWT format.")
        return

    header_b64, payload_b64, signature_b64 = parts

    # Decode header
    # Add padding if necessary
    padded = header_b64 + '=' * (-len(header_b64) % 4)
    header_json_str = base64.urlsafe_b64decode(padded).decode('utf-8')
    alg = extract_alg_from_header(header_json_str)
    if not alg:
        print("Failed to extract alg from header.")
        return

    signing_input = f"{header_b64}.{payload_b64}"

    # Start timer
    start_time = time.time()
    t = threading.Thread(target=timer_thread)
    t.daemon = True
    t.start()

    # Brute-force
    for length in range(1, max_length + 1):
        if brute_force(signing_input, signature_b64, alg, length, 0, [''] * length):
            running = False
            total = time.time() - start_time
            print(f"\n✅ Found secret!   : {candidate:<30} Total time  : {total:.2f} seconds")
            return

    running = False
    total = time.time() - start_time
    print(f"\n❌ Secret not found. Total time  : {total:.2f} seconds")


if __name__ == '__main__':
    main()

File: JwtBruteForce.java
import java.nio.charset.StandardCharsets;
import java.util.Base64;
import java.util.Scanner;

import javax.crypto.Mac;
import javax.crypto.spec.SecretKeySpec;

public class JwtBruteForce {
  private static String candidate = "";
  private static long startTime = 0;
  private static final char[] CHARSET = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789!@#$%^&*()-_+=[]{};:'\",.<>?/\\|`~".toCharArray();

  public static void main(String[] args) throws Exception {
    Scanner sc = new Scanner(System.in);
    System.out.print("Enter JWT token\t\t\t: ");
    String jwt = sc.nextLine().trim();

    System.out.print("Enter max secret length to try\t: ");
    int maxLength = Integer.parseInt(sc.nextLine().trim());
    sc.close();

    String[] parts = jwt.split("\\.");
    if (parts.length != 3) {
      System.out.println("Invalid JWT format.");
      return;
    }

    String headerB64 = parts[0];
    String payloadB64 = parts[1];
    String signatureB64 = parts[2];

    String headerJson = new String(Base64.getUrlDecoder().decode(headerB64), StandardCharsets.UTF_8);
    String alg = extractAlgFromHeader(headerJson);
    if (alg == null) {
      System.out.println("Failed to extract alg from header.");
      return;
    }

    String javaAlg = switch (alg) {
      case "HS256" -> "HmacSHA256";
      case "HS384" -> "HmacSHA384";
      case "HS512" -> "HmacSHA512";
      default -> throw new IllegalArgumentException("Unsupported alg: " + alg);
    };

    String signingInput = headerB64 + "." + payloadB64;

    startTime = System.nanoTime();

    for (int length = 1; length <= maxLength; length++) {
      if (bruteForce(signingInput, signatureB64, javaAlg, new char[length], 0, startTime)) {
        double total = (System.nanoTime() - startTime) / 1e9;
        System.out.printf("\n✅ Found secret!\t: %-30s Total time\t: %.2f seconds%n", candidate, total);
        return;
      }
    }

    double total = (System.nanoTime() - startTime) / 1e9;
    System.out.printf("\n❌ Secret not found. Total time: %.2f seconds%n", total);
  }

  private static boolean bruteForce(String signingInput, String targetSig, String alg, char[] current, int pos, long startTime) throws Exception {
    long elapsed = System.nanoTime() - startTime;
    long millis = elapsed / 1_000_000;
    long hundredths = (millis % 1000) / 10;
    long seconds = (millis / 1000) % 60;
    long minutes = (millis / (1000 * 60)) % 60;
    long hours = (millis / (1000 * 60 * 60));

    System.out.printf("⏱  Trying\t\t: %-30s Time elapsed\t: %03d:%02d:%02d.%02d\r", candidate, hours, minutes, seconds, hundredths);

    if (pos == current.length) {
      candidate = new String(current);
      String testSig = hmacBase64Url(signingInput, candidate, alg);
      return testSig.equals(targetSig);
    }

    for (char c : CHARSET) {
      current[pos] = c;
      if (bruteForce(signingInput, targetSig, alg, current, pos + 1, startTime)) {
        return true;
      }
    }

    return false;
  }

  private static String extractAlgFromHeader(String headerJson) {
    int algIndex = headerJson.indexOf("\"alg\"");
    if (algIndex == -1) return null;

    int colon = headerJson.indexOf(':', algIndex);
    int quote1 = headerJson.indexOf('"', colon + 1);
    int quote2 = headerJson.indexOf('"', quote1 + 1);
    return headerJson.substring(quote1 + 1, quote2);
  }

  private static String hmacBase64Url(String data, String key, String algorithm) throws Exception {
    Mac mac = Mac.getInstance(algorithm);
    SecretKeySpec secretKey = new SecretKeySpec(key.getBytes(StandardCharsets.UTF_8), algorithm);
    mac.init(secretKey);
    byte[] hmac = mac.doFinal(data.getBytes(StandardCharsets.UTF_8));
    return Base64.getUrlEncoder().withoutPadding().encodeToString(hmac);
  }
}
